Parse and pass streamlookup parameters

yumkam opened the pull request:
### Changelog entry <!-- a user-readable short description of changes introduced in this PR -->

...

### Changelog category <!-- remove all except one -->

* Improvement
* Not for changelog (changelog entry is not required)

### Additional information

...

yumkam committed:
change DqJoin Flags from TCoAtomList into TCoNameValueTupleList

yumkam committed:
pass streamlookup() join parameters

yumkam committed:
Reject unsupported streamlookup variants

yumkam committed:
fix rebase error

yumkam committed:
fix typo

yumkam commented:
Changed type for DqJoin and GraceJoinCore Flags from TCoAtomList to TCoNameValueTupleList
(this will break canonized tests)

yumkam committed:
fix ut

yumkam committed:
fix index out of range

yumkam committed:
fix typo

yumkam committed:
fix streamlookup sql to match new requirements

yumkam committed:
rephrase error message

rvu1024 commented on ydb/library/yql/dq/opt/dq_opt_join.cpp:
```diff
@@ -522,20 +544,26 @@ TExprBase DqRewriteRightJoinToLeft(const TExprBase node, TExprContext& ctx) {
+                list.push_back(Build<TCoNameValueTuple>(ctx, flag.Pos())
+                        .Name().Build("Right" + TString(tail))
+                    .Done());
+            } else if ( flag.Name().Value().AfterPrefix("Right", tail)) {
```
style: `if (flag`

rvu1024 commented on ydb/library/yql/dq/type_ann/dq_type_ann.cpp:
```diff
@@ -501,15 +502,35 @@ const TStructExprType* GetDqJoinResultType(const TExprNode::TPtr& input, bool st
+    if (input->ChildrenSize() > TDqJoin::idx_Flags) {
+        auto&& flags = input->Tail();
+        for (ui32 i = 0; i < flags.ChildrenSize(); ++i) {
+            auto&& flag = *flags.Child(i);
```
Зачемм тут &&?

rvu1024 commented on ydb/library/yql/dq/type_ann/dq_type_ann.cpp:
```diff
@@ -501,15 +502,35 @@ const TStructExprType* GetDqJoinResultType(const TExprNode::TPtr& input, bool st
+        auto&& flags = input->Tail();
+        for (ui32 i = 0; i < flags.ChildrenSize(); ++i) {
+            auto&& flag = *flags.Child(i);
+            if ( !EnsureTupleOfAtoms(flag, ctx) || !EnsureTupleMinSize(flag, 1, ctx))
```
style: `if (!`

rvu1024 commented on ydb/library/yql/dq/type_ann/dq_type_ann.cpp:
```diff
@@ -501,15 +502,35 @@ const TStructExprType* GetDqJoinResultType(const TExprNode::TPtr& input, bool st
+            auto&& flag = *flags.Child(i);
+            if ( !EnsureTupleOfAtoms(flag, ctx) || !EnsureTupleMinSize(flag, 1, ctx))
                 return nullptr;
+            auto&& name = *flag.Child(TCoNameValueTuple::idx_Name);
```
зачем тут &&?

yumkam replied:
`auto` нельзя, `const auto&` нельзя, остаётся выбор между `auto&` и `auto&&`

rvu1024 commented on ydb/library/yql/dq/type_ann/dq_type_ann.cpp:
```diff
@@ -501,15 +502,35 @@ const TStructExprType* GetDqJoinResultType(const TExprNode::TPtr& input, bool st
+                       return nullptr;
+                   continue;
+                }
+            } else
+            if (name.IsAtom({"LeftAny", "RightAny"})) {
```
style: лишний перенос

rvu1024 commented on ydb/library/yql/providers/dq/opt/physical_optimize.cpp:
```diff
@@ -252,6 +252,38 @@ class TDqsPhysicalOptProposalTransformer : public TOptimizeTransformerBase {
+            for (auto&& flag: maybeFlags.Cast()) {
+                auto&& name = flag.Name().Value();
+                if (name == "TTL"sv) {
+                    YQL_ENSURE(flag.Ref().ChildrenSize() == 2);
+                    ttl = flag.Ref().Child(1);
+                } else if (name == "MaxCachedRows"sv) {
+                    YQL_ENSURE(flag.Ref().ChildrenSize() == 2);
+                    maxCachedRows = flag.Ref().Child(1);
+                } else if (name == "MaxDelayedRows"sv) {
+                    YQL_ENSURE(flag.Ref().ChildrenSize() == 2);
+                    maxDelayedRows = flag.Ref().Child(1);
+                } else if (name == "RightAny"sv) {
+                    YQL_ENSURE(flag.Ref().ChildrenSize() == 1);
```
Все YQL_ENSURE должны быть проверками в типизаторе. Здесь их уже не должно быть.
И там же, в типизаторе нужно провверить, что в настроках лежат числа и они в пределах допустимых лимитов

rvu1024 commented on ydb/library/yql/providers/dq/opt/physical_optimize.cpp:
```diff
@@ -252,6 +252,38 @@ class TDqsPhysicalOptProposalTransformer : public TOptimizeTransformerBase {
+                    YQL_ENSURE(flag.Ref().ChildrenSize() == 1);
+                    continue;
+                } else {
+                    Y_DEBUG_ABORT("Impossible flag"); // should have been checked by dq_type_ann
```
ctx.AddError + return {}

rvu1024 reviewed (changes requested):
Все Child(0) можно заменить на Head()

yumkam committed:
tests/fq/generic/streaming: update join to ANY

yumkam committed:
code style fixes

yumkam committed:
move validation to correct place

yumkam committed:
Add test with parameters

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

Conflicts:
	ydb/library/yql/providers/dq/planner/execution_planner.cpp

yumkam commented on ydb/library/yql/core/yql_join.cpp:
```diff
@@ -319,8 +319,16 @@ namespace {
-                if (!EnsureTupleSize(*child, 1, ctx)) {
-                    return IGraphTransformer::TStatus::Error;
+                if (option.IsAtom("forceStreamLookup")) {
+                    if (child->ChildrenSize() % 2 == 0) {
```
До EquiJoin список as is передаётся внутри `('forceStreamLookup [...])` как список. При трансляции в DqJoin список разбивается на пары передаётся в Flags (которые переделан со списка Atom на список NameValueTuple.
В аннотаторе типов DqJoin проверяется имена флагов на допустимость.
При трансформации в DqJoin в CnStreamLookup проверяется, что слева нет Any и справа есть Any (в CnStreamLookup нет произвольных флагов).
В аннотаторе типов CnStreamLookup проверяется, что Join -- LEFT, и значения параметров -- целые числа.

yumkam commented:
(Resolution of trivial conflict, tests canonization and adding test for parameters is delayed till final iteration)

yumkam committed:
fix typo in new test

yumkam committed:
recanonize tests

yumkam committed:
Style fixes

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
Fix merge conflict with #8169

yumkam committed:
fixup! Fix merge conflict with #8169

yumkam committed:
style fixes

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

rvu1024 commented on ydb/library/yql/dq/opt/dq_opt_join.cpp:
```diff
@@ -566,20 +582,26 @@ TExprBase DqRewriteRightJoinToLeft(const TExprBase node, TExprContext& ctx) {
-            } else if ( flag.Value().AfterPrefix("Right", tail)) {
-                flagsBuilder.Add().Value("Left" + TString(tail)).Build();
+            if (flag.Name().Value().AfterPrefix("Left", tail)) {
+                list.push_back(Build<TCoNameValueTuple>(ctx, flag.Pos())
+                        .Name().Build("Right" + TString(tail))
+                    .Done());
+            } else if (flag.Name().Value().AfterPrefix("Right", tail)) {
+                list.push_back(Build<TCoNameValueTuple>(ctx, flag.Pos())
+                        .Name().Build("Left" + TString(tail))
+                    .Done());
```
.InitFrom(flag)

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
review fixes

DqRewriteRightJoinToLeft: copy tuples and rewrite names

yumkam committed:
review fixes (IntFromString -> FromString)

maximyurchuk reviewed (changes requested):
YQL правится в Аркадии, в gh же просто зеркало, надо нести правки в Аркадию

yumkam commented:
> YQL правится в Аркадии, в gh же просто зеркало, надо нести правки в Аркадию

Тут взаимозависимые изменения и в yql/ и в ydb/

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
streamlookup: temporary weaken ANY requirement

Currently LEFT JOIN ANY rejected, so for transition weaken requirements
to warning

yumkam committed:
add forgotten file

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
(re)canonize

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
Merge remote-tracking branch 'origin/main' into pass-in-inputtransform-maxdelay

yumkam committed:
update canonization

yumkam committed:
fixup

yumkam commented:
gh часть будет решена в https://github.com/ydb-platform/ydb/pull/12548
